[ENH] Add a log_likelihood method for DynamicBayesianNetwork

petrello opened the pull request:
### Your checklist for this pull request
- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2228

### List of changes to the codebase in this pull request
- Adds `log_likelihood(self, data)` method to `DynamicBayesianNetwork` class
- Adds util method `_log_likelihood_node(self, node, sample_data, cpd)` to `DynamicBayesianNetwork` class
- Adds new class of unit tests `TestLogLikelihood`

petrello committed:
Add log_likelihood definition to DBN class

petrello committed:
Add doc and example for log_likelihood function

petrello committed:
Fix: F841 local variable 'pbar' is assigned to but never used

petrello committed:
Fix: F401 'pgmpy.sampling.BayesianModelSampling' imported but unused

petrello committed:
Implement log_likelihood method for DynamicBayesianNetwork

- Added input validation
- Improved function doc-string
- Added function implementation

petrello committed:
Update docstring for log_likelihood method in DynamicBayesianNetwork

Added show_progress parameter in the score method's docstring for improved clarity.

petrello committed:
Add node log-likelihood computation in DBN

Implement _log_likelihood_node method for calculating node log-probabilities in DynamicBayesianNetwork.

petrello committed:
Replace _nodes() with nodes() in log_likelihood method

Updated references from the internal _nodes() method to the public nodes() method in the log_likelihood function ensure comparison with dataframe columns

petrello committed:
Fix sample_data key access in DynamicBayesianNetwork

Updated key access in sample_data to use (node[0], node[1]) tuple format for both nodes and parents. This ensures correct retrieval of observed and parent values when working with multi-indexed sample_data.

petrello committed:
Improve log_likelihood docstrings and input validation

Expanded and clarified docstrings for log_likelihood and its helper method, including detailed parameter, return, and error descriptions. Improved input validation and error messages for data structure and content. Updated usage examples for clarity and accuracy.

petrello committed:
Add tests for DBN log_likelihood method

Introduces a new TestLogLikelihood class to test the log_likelihood method of DynamicBayesianNetwork. Tests include input validation, consistency of results, and comparison of log-likelihoods for normal versus anomalous data.

petrello committed:
Reformat tqdm call for improved readability

Adjusted the formatting of the tqdm call in the log-likelihood computation loop to improve code readability by placing each argument on a separate line.

dbolser commented:
I think I've fixed the black issues. You just need to merge my PR on your PR @petrello :-)

petrello committed:
fix: reformatting file DynamicBayesianNetwork

fix code format according to Black suggestions

petrello committed:
fix: reformatting DynamicBayesianNetwork

- fixed other sections of code in the DynamicBayesianNetowrk class according to Balck suggestions

petrello committed:
fix: black reformatting test_DynamicBayesianNetowrk

- fixes some code sections style of class test_DynamicBayesianNetwork according to Black suggestions.

petrello committed:
fix: linting corrections for test_DynamicBayesianNetwork

petrello committed:
fix doc formatting in DynamicBayesianNetwork

- fix the doc formatting for the log_likelihood function of the DynamicBayesianNetowrk class

petrello committed:
fix docstring in log_likelihood

petrello committed:
Update DynamicBayesianNetwork.py

petrello committed:
Update DynamicBayesianNetwork.py

dbolser commented:
Thanks

petrello commented:
@dbolser Sorry for the late reply, and thanks a lot for your PR. I decided not to merge it, as I opted to fix the formatting manually to keep this PR self-contained.

Regarding the Black configuration issue, I believe it would be more appropriate to open a separate issue and then wait for feedback from @ankurankan before proceeding with a fix.

Thanks again!

dbolser commented:
No problem, are the black CI tests still failing?

petrello commented:
Not for this PR because I manually formatted the code to match Black's requirements. However, it looks like the underlying Black CI issue is still present in the main pgmpy repository.

ankurankan commented:
@petrello @dbolser Thanks for the PR and sorry for the super late reply. Is this ready for a review? 

For the formatting issues, we have a pre-commit hook (https://github.com/pgmpy/pgmpy/blob/dev/.pre-commit-config.yaml) that you can install, and it will take care of all formatting fixes automatically. You just need to run the following from the pgmpy base directory:

```
pip install pre-commit
pre-commit install
```

petrello commented:
Hi @ankurankan, no worries at all. Yes, the PR is ready to be reviewed.

Also, thank you for the tip about the pre-commit hook. I might have missed some steps while configuring my local environment.

dbolser commented:
Cool, the commit hook sounds good!

You can close this now (I think).

ankurankan commented:
@dbolser, I think all the tests are passing, so no need to make any changes (or close this) in this PR regarding the hook. And sorry again for the delay in the review, I still haven't gotten a chance to go through it.

@XAheli @Spinachboul @Nimish-4 Would any of you be interested in doing a first round of reviews?

XAheli commented:
> @dbolser, I think all the tests are passing, so no need to make any changes (or close this) in this PR regarding the hook. And sorry again for the delay in the review, I still haven't gotten a chance to go through it.
> 
> @XAheli @Spinachboul @Nimish-4 Would any of you be interested in doing a first round of reviews?

Surely!!

Spinachboul commented:
> @dbolser, I think all the tests are passing, so no need to make any changes (or close this) in this PR regarding the hook. And sorry again for the delay in the review, I still haven't gotten a chance to go through it.
> 
> @XAheli @Spinachboul @Nimish-4 Would any of you be interested in doing a first round of reviews?


Yes, will do!

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+                f"Data must be a pandas.DataFrame instance. Got: {type(data)}"
+            )
+
+        if min(data.columns, key=lambda x: x[1])[1] != 0:
+            raise ValueError("Data column names must start from time slice 0.")
+
```
What happens if `data.columns` is empty? The `min()` function will raise a `ValueError`. Should we add a check for empty DataFrames before this validation? 

also what if columns are not tuples at all? This will cause a `TypeError` when accessing `x[1]`. Should tuple structure validation happen prior this check?

petrello replied:
I propose to check if `data` is empty or has empty columns (all) using `data.dropna().empty` property. This way, we ensure `data` has at least one non-empty column so that `min()` won't fail. Then, check for each `data.columns` if the column name is a tuple and has a length of 2, and whether the second item of the tuple is a non-negative integer.

```suggestion
        if data.dropna().empty:
            raise ValueError("DataFrame is empty.")

        for col in data.columns:
            if not isinstance(col, tuple) or len(col) != 2:
                raise ValueError(
                    "Data column names must be tuples of the form (node_name, time_slice). "
                    f"Got column type {type(col)} of length {len(col) if isinstance(col, tuple) else 'N/A'}"
                )
            if not isinstance(col[1], int) or col[1] < 0:
                raise ValueError(
                    "Time slice in data column names must be a non-negative integer."
                )
        
        if min(data.columns, key=lambda x: x[1])[1] != 0:
            raise ValueError("Data column names must start from time slice 0.")
```

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+        if min(data.columns, key=lambda x: x[1])[1] != 0:
+            raise ValueError("Data column names must start from time slice 0.")
+
+        if set(self.nodes()) - set(data.columns):
+            raise ValueError(
+                f"Missing columns in data. Can't find values for the following variables: "
+                f" {set(self.nodes()) - set(data.columns)}"
+            )
```
Should include a validation that `data.columns` is a subset of or exactly matches `self.nodes()`

petrello replied:
Yes, we can add the validation step you mentioned, but it might not be needed. Notice that extra columns in `data.columns` won't be considered since the method will refer only to `self.nodes()`. Therefore, we are only interested in checking if the dataframe `data` contains the nodes.

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+        self.check_model()
+
+        # Create mapping from node to CPD for efficient lookup
+        node_to_cpd = {cpd.variable: cpd for cpd in self.get_cpds()}
```
`node_to_cpd` is built from `self.get_cpds()` but `get_cpds()` could return empty and fail or raise KeyError. Given that users may call `log_likelihood` before fitting/adding cpds, error should be clear

petrello replied:
Ok, we can either validate the result of `self.get_cpds` or the dictionary `node_to_cpd`. For example, if `get_cpds()` returns empty, then `note_to_cpd` will be empty as well.

We can opt for something like:
```suggestion
        # Create mapping from node to CPD for efficient lookup
        node_to_cpd = {cpd.variable: cpd for cpd in self.get_cpds()}

        if not node_to_cpd:
            raise ValueError("No CPDs found in the model.")
```

To manage possible KeyError exceptions, we can define a `cpd` variable and wrap it with a `try-except` statement:
```python
    # Compute log-probability for each node in this time slice
    for node in nodes_t:
        try:
            cpd = node_to_cpd[node]
        except:
            raise KeyError(
                f"CPD for node {node} not found in the model."
            )

        node_log_prob = self._log_likelihood_node(
            node, data.iloc[i], cpd
        )
        sample_log_likelihood += node_log_prob
```
since `node_to_cpd[node]` could raise that kind of exception.

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+
+            # Process each time slice
+            for t in sorted(time_slices):
+                # Get nodes in current time slice
+                nodes_t = [node for node in nodes if node[1] == t]
+
+                # Compute log-probability for each node in this time slice
+                for node in nodes_t:
+                    node_log_prob = self._log_likelihood_node(
+                        node, data.iloc[i], node_to_cpd[node]
+                    )
+                    sample_log_likelihood += node_log_prob
```
The docstring mentions "Variables in the first time slice (t=0) use their marginal probabilities" but I don't see handling for t=0 vs t>0. Is this handled implicitly by the parent structure?

cc @ankurankan

petrello replied:
The decision of using marginal probability rather than conditional probability is handled inside the `_log_likelihood_node` method as follows:

```python
# Get parents (evidence variables) for this node
parents = self.get_parents(node)

if not parents:
    # No parents - use marginal probability
    ...
else:
    # Has parents - use conditional probability
    ...
```

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+
+        # Ensure probability is not zero for numerical stability
+        # (with a small epsilon to avoid log(0))
+        prob = max(prob, 1e-10)
```
Why specifically 1e-10 as the epsilon value? Is this documented anywhere?

petrello replied:
No, this is actually just a small value to avoid log(0).

XAheli commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+                "requires complete observations."
+            )
+
+        self.check_model()
```
maybe adding a parameter to `log_likelihood` to allow skipping model validation when we know the model is unchanged ?

XAheli commented on pgmpy/tests/test_models/test_DynamicBayesianNetwork.py:
more edge cases would be great
e.g. 

- `test_log_likelihood_empty_df_raises`
- `test_log_likelihood_invalid_column_names_raises`
- `test_log_likelihood_extra_columns_warns`
- `test_log_likelihood_no_cpds_raises`
- `test_log_likelihood_epsilon_warning`

petrello replied:
Perfect, thank you.

What do you mean by `test_log_likelihood_epsilon_warning`?

XAheli reviewed (commented):
Added some comments kindly look into it.

Spinachboul commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+
+        if not parents:
+            # No parents - use marginal probability
+            prob = cpd.values[observed_state_idx]
```
A shape check here will probably make sense for `cpd_values` for correct indexing else could return `IndexError` or other incompatibility issues

Spinachboul commented on pgmpy/models/DynamicBayesianNetwork.py:
```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+        """
+
+        # Get the observed value for this node
+        observed_value = sample_data[(node[0], node[1])]
```
Could we encounter `NaN` values here, and handling them gracefully makes sense?
Also if the node key `node` does not exist in the `sample_data.index`, then what happens?

dbolser commented:
Just to be clear, this is @petrello's PR. I just injected myself for some practice.

petrello committed:
feat: add new tests for DBN loglikelihood

- test if input dataframe has no data
- test if input dataframe has wrong column format
- test if warns when input dataframe has extra column

petrello committed:
feat: add validation steps to DBN loglikelihood

- check if empty dataframe in input
- check correct column format in dataframe
- check for extra column in dataframe
- check if model has no CPDs
- more robust cpd retrieval

dbolser commented:
I'll leave this to you guys

petrello commented:
@ankurankan @Spinachboul @XAheli Hi! Just checking back on this PR after the latest updates.

I've addressed the previous points and refined the implementation of `log_likelihood` for `DynamicBayesianNetwork`, aiming to keep it consistent with existing APIs.

I'd really appreciate any feedback when you have time. Happy to make further changes if needed!
